xla/hlo/transforms/collectives/all_gather_cse.cc

```diff
@@ -0,0 +1,119 @@
+/* Copyright 2024 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/hlo/transforms/collectives/all_gather_cse.h"
+
+#include "xla/hlo/ir/hlo_instruction.h"
+#include "xla/hlo/ir/hlo_opcode.h"
+#include "xla/hlo/transforms/simplifiers/hlo_dce.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/statusor.h"
+
+namespace xla {
+
+absl::StatusOr<bool> AllGatherCSE::Run(
+    HloModule* module,
+    const absl::flat_hash_set<absl::string_view>& execution_threads) {
+  VLOG(2) << "Running AllGatherCSE pass";
+  bool changed = false;
+
+  for (HloComputation* computation : module->computations(execution_threads)) {
+    VLOG(2) << "Processing computation: " << computation->name();
+    absl::flat_hash_map<std::tuple<HloInstruction*, int64_t, PrimitiveType>,
+                        HloInstruction*>
+        all_gather_map;  // Every region has its own all gather map
+    for (HloInstruction* instruction : computation->instructions()) {
+      if (instruction->opcode() == HloOpcode::kAllGather) {
+        VLOG(2) << "Found all-gather instruction: " << instruction->ToString();
+        auto raw_parameter_tuple =
+            FindRawParameter(instruction->mutable_operand(0));
+        HloInstruction* raw_parameter = std::get<0>(raw_parameter_tuple);
+        if (raw_parameter != nullptr &&
+            raw_parameter->opcode() == HloOpcode::kParameter) {
+          auto it = all_gather_map.find(raw_parameter_tuple);
+          if (it != all_gather_map.end()) {
+            VLOG(2) << "Replacing all-gather with previous result: "
+                    << it->second->ToString();
+            TF_RETURN_IF_ERROR(instruction->ReplaceAllUsesWith(it->second));
+            changed = true;
+          } else {
+            VLOG(2) << "Storing all-gather result for future use";
+            all_gather_map[raw_parameter_tuple] = instruction;
+          }
+        }
+      }
+    }
+  }
+  if (changed) {
+    // Run DCE if changed.
+    HloDCE dce;
+    TF_RETURN_IF_ERROR(dce.Run(module).status());
```

Comment:
Should be a following pass in the surrounding pipeline

Reply:
Do you recommend I remove dead instructions with TF_REMOVE_INSTRUCTION_FROM_COMPUTATION?

Or should I just run DCE in my unit tests?

Reply:
Resolved to run DCE in unit tests. Removed nested DCE from pass.